pyonfx/timestamps.py

```diff
@@ -0,0 +1,552 @@
+                timestamps.append(rounding_method(time))
+                time += Fraction(1000) / range_v1.fps
+                frame += 1
+
+        timestamps.append(rounding_method(time))
+        fpms = default_fps / Fraction(1000)
+        return timestamps, time, fpms
+
+    def _parse_v2_and_v4_file(
+        file_content: TextIOWrapper, version: int, rounding_method: RoundingMethod
+    ) -> Tuple[List[int], Fraction, Fraction]:
+        """Create timestamps based on the timestamps v1 file provided.
+
+        Inspired by: https://gitlab.com/mbunkus/mkvtoolnix/-/blob/72dfe260effcbd0e7d7cf6998c12bb35308c004f/src/merge/timestamp_factory.cpp#L201-267
+
+        Parameters:
+            file_content (TextIOWrapper): The timestamps content
+            version (int): The version of the timestamps (only 2 or 4 is allowed)
+            rounding_method (RoundingMethod): A rounding method
+
+        Returns:
+            A tuple containing these 3 informations:
+                1. A list of each timestamps rounded to milliseconds
+                2. The last timestamps not rounded
+                3. The fpms
+        """
+
+        if version != 2 and version != 4:
+            raise ValueError("You can only specify version 2 or 4.")
+
+        timestamps: List[int] = []
+        previous_timestamp: int = 0
+        lowest_timestamp: Fraction = None
+        highest_timestamp: Fraction = None
+
+        for line in file_content:
+            line = line.strip(" \t")
+
+            if not line or line.startswith("#"):
+                continue
+
+            try:
+                timestamp = Fraction(line)
+            except ValueError:
+                raise ValueError(
+                    f'The timestamps file contain a invalid line. Here is it: "{line}"'
+                )
+
+            if highest_timestamp is None or highest_timestamp < timestamp:
+                highest_timestamp = timestamp
+            if lowest_timestamp is None or lowest_timestamp > timestamp:
+                lowest_timestamp = timestamp
+
+            rounded_timestamp = rounding_method(timestamp)
+
+            if version == 2 and rounded_timestamp < previous_timestamp:
+                raise ValueError(
+                    f"The timestamps file contain timestamps NOT in ascending order."
+                )
+
+            previous_timestamp = rounded_timestamp
+            timestamps.append(rounded_timestamp)
+
+        if not len(timestamps):
+            raise ValueError(f"The timestamps file is empty.")
+
+        if version == 4:
+            timestamps.sort()
+
+        duration = highest_timestamp - lowest_timestamp
+        if duration:
+            fpms = Fraction(len(timestamps) - 1) / (
+                highest_timestamp - lowest_timestamp
+            )
+        else:
+            fpms = 0
+        return timestamps, highest_timestamp, fpms
+
+
+class Timestamps:
+    """Timestamps object contains informations about the timestamps of an video.
```

Comment:
Moreover, users reading documentation need to know that they need to call one of the factory methods (from_fps, from_timestamps_file, from_video_file).

Reply:
Does this sound good to you?
```py
"""Timestamps object contains informations about the timestamps of an video.
Both Constant Frame Rate (CFR) and Variable Frame Rate (VFR) videos are supported.

To create an Timestamps object, you need to call one of the following method ``from_fps``, ``from_timestamps_file``, ``from_video_file``.
Ex:
	timestamps = Timestamps.from_fps(Fraction(24000, 1001))


Parameters:
	rounding_method (RoundingMethod): A rounding method. 99% of the time, you want to use RoundingMethod.ROUND.
		Note:
			Video player have 2 methods to deal with timestamps. Some floor them and other round them.
			This can lead to difference when displaying the subtitle.
				Ex:
					Player - Method - proof
					mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
					FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L194-L196
					VLC    - floor  - https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/modules/codec/libass.c#L453-454
									  https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L120-132
					MPC-HC - floor  - https://github.com/clsid2/mpc-hc/blob/0994fd605a9fb4d15806d0efdd6399ba1bc5f984/src/Subtitles/LibassContext.cpp#L843
			Important note:
				Matroska (.mkv) file are an exception !!!
				If you want to be compatible with mkv, use RoundingMethod.ROUND.
				By default, they only have a precision to milliseconds instead of nanoseconds like most format.
					For more detail see:
						1- https://mkvtoolnix.download/doc/mkvmerge.html#mkvmerge.description.timestamp_scale
						2- https://matroska.org/technical/notes.html#timestampscale-rounding
	timestamps (List[int], optional): A list of [timestamps](https://en.wikipedia.org/wiki/Timestamp) in milliseconds encoded as integers.
							It represent each frame [presentation timestamp (PTS)](https://en.wikipedia.org/wiki/Presentation_timestamp)
	normalize (bool, optional): If True, it will shift the timestamps to make them start from 0. If false, the option does nothing.
	fpms (Fraction, optional): The fpms.
	last_frame_time (Fraction, optional): The last frame time not rounded.
"""
```

Reply:
Better, but it's still not correct, because you created a more general method (and not only one that "contains information about the timestamps of a video").

Reply:
Could you update the doc for this?
I don't understand what you want

Reply:
Sure, I will do it later